src/compiler/transformers/ts.ts

```diff
@@ -1242,10 +1247,12 @@ namespace ts {
          */
         function addInitializedPropertyStatements(statements: Statement[], properties: ReadonlyArray<PropertyDeclaration>, receiver: LeftHandSideExpression) {
             for (const property of properties) {
-                const statement = createExpressionStatement(transformInitializedProperty(property, receiver));
+                if (isPrivateProperty(property)) {
+                    continue;
+                }
+                const statement = createStatement(transformInitializedProperty(property, receiver));

```

Comment:
By bailing out early here, it changes the order in which the property initializers are evaluated. We need to preserve a strict top-down evaluation order. Probably the right way to do this is to add the statements for non-private fields into the same `statements` array being added to below, preserving order.

Reply:
Input:

    let a = 0;
    class C {
        foo = ++a;
        #x = ++a;
    }

Output:

    let a = 0;
    class C {
        constructor() {
            _x.set(this, ++a); // BAD: evaluation order is swapped
            this.foo = ++a; // BAD: evaluation order is swapped
        }
    }
    var _x = new WeakMap;

Reply:
This is somewhat complicated because the TypeScript transformer will transform class property initializers to constructor statements. If the target output is ESNext, then we need to keep the private named property declaration in the class to keep valid ESNext syntax.

After further thought, I now think that we should move this behavior for ALL properties (not just private names) to the ESNext transformer, since you can now have property initializers in ESNext. (I'm also hoping that the initializer semantics are the same in the ESNext proposal as they are in the TypeScript class implementation). I'd like to get the TypeScript team's thoughts on this idea.

Reply:
I've now fixed this issue with private-named property initializers. It now relies on the TS transformer moving the initializers to the constructor (in the original order) and initializing the WeakMap values to undefined at the top of the constructor (like `_x.set(this, void 0);`).